### PR TITLE
ACC-197 new modify email rpc endpoint

### PR DESCRIPTION
**Pull request by esclapes**
This PR adds a new RPC method to `updateEmail` of a user. The user password is required to update the email.

## Thing to note:

* ~First RPC method that I am adding, do I need to register the method anywhere other that `/rpc/index.js` and the main `indfex.js`?~ clear I tested the endpoint manually
* This will be discussed in a differet PR between cycles. ~I wanted to give it a go at using [this approach for testing the mongoose models](https://jestjs.io/docs/en/mongodb). It works locally, but [there is an issue running them inside alpine](https://github.com/nodkz/mongodb-memory-server/issues/32). If we are running the tests in CI, __this will break__ I am happy to remove this part and add some tests with mocks. How do you feel about it?~

### CODE REVIEW

**Comment by esclapes** [comment]
@philippemiguet  @hharnisc I would like to add some extra tests for parameter validation on the new method. Let me know which approach you find more appropriate for tests at the moment, and I will add them.

File: src/rpc/updateEmail.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+const { createError } = require('@bufferapp/micro-rpc')
+const AuthenticationAccountModel = require('../models/authenticationAccountModel')
+
+module.exports = async ({ _id, email, password, newEmail }) => {
```
Comment (philippemiguet) [inline src/rpc/updateEmail.js]: @esclapes - I see we are expecting a password here to enforce security, this is a great idea! One thing, you forgot to use it later in the function to authenticate the user :)

File: tests/rpc/updateEmail.test.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
+const mongoose = require('mongoose')
+const { updateEmail } = require('../../src/rpc')
+const AuthenticationAccountModel = require('../../src/models/authenticationAccountModel')
+
+jest.unmock('../../src/models/authenticationAccountModel')
```
Comment (philippemiguet) [inline tests/rpc/updateEmail.test.js]: Should we still mock the model if we use in-memory mongo?

File: tests/rpc/updateEmail.test.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+describe('updateEmail', () => {
```
Comment (philippemiguet) [inline tests/rpc/updateEmail.test.js]: Great start here too! Would you be up adding one test per condition, i.e. `if(!password)`, in the updateEmail.js file? :)

File: tests/rpc/updateEmail.test.js (hunk 2 above)
Comment (esclapes) [reply tests/rpc/updateEmail.test.js]: It is actually _unmocking_ here, but you bring up a good point. 

I saw there is a `__mocks__` directory next to our model, and I wrongly thought that it meant I had to _unmock_ it here. However, [reading the docs again](https://jestjs.io/docs/en/manual-mocks.html), it seems to not be necessary. I will remove it.

Good remark @philippemiguet

File: tests/rpc/updateEmail.test.js (hunk 3 above)
Comment (esclapes) [reply tests/rpc/updateEmail.test.js]: Sure, it is a good practice to test that validations are preventing the update.

**Comment by esclapes** [comment]
@philippemiguet do you have access to see the details on this check: `k8sjenkins/cicd — k8sjenkins: Build failed`

File: src/rpc/updateEmail.js (hunk 1 above)
Comment (esclapes) [reply src/rpc/updateEmail.js]: ah! Great one @philippemiguet. I need to fix that.

**Comment by philippemiguet** [comment]
> do you have access to see the details on this check: k8sjenkins/cicd — k8sjenkins: Build failed

@esclapes - Yes I do. If you login to Jenkins using Okta you should have access as well. Regarding the issue itself, Adnan and Harrison are working on this, there is an issue with Helm chart, no need to worry about this :)

File: tests/rpc/updateEmail.test.js (hunk 2 above)
Comment (philippemiguet) [reply tests/rpc/updateEmail.test.js]: Haha, I didn't even notice we were "unmocking" 🤦‍♂️ Good my comment was still useful 😂

File: src/rpc/updateEmail.js (hunk 1 above)
Comment (esclapes) [reply src/rpc/updateEmail.js]: I have pushed a change to fix this @philippemiguet 

Still some tests missing. Waiting on @hharnisc to review the in-memory-mongo approach to testing and see if it makes sense to use it or I should rather change the approach to using mocks.

**Comment by esclapes** [comment]
@philippemiguet As discussed, I have removed the in-memory mongodb part and used regular mocking for testing the new RPC method. 

This way we can decouple the decision about testing from this new method which is required by the reply team. cc @CoxyBoris

**Comment by esclapes** [comment]
By the way @philippemiguet we may want to squash & merge this one.

**Comment by djfarrelly** [comment]
@esclapes @philippemiguet I see the comment about alpine and whatnot, what exactly is the issue there? I'm unclear on that 😄 

We've used Alpine in production in an effort to keep our production containers as small as possible. This should enable faster builds (when pulling the base image) and much faster startup times when a Kubernetes node needs to pull the image to start the service. Does that help explain things? Let me know if y'all have follow up questions! 

PS - We do use non-Alpine images in production for some services that require more libraries and whatnot. We're not stuck on using Alpine, but we'd aim to use it wherever possible.

**Comment by esclapes** [comment]
Hey @djfarrelly! Thanks for checking out this thread. The `node:alpine` image works fine.

I did discover an incompatibility issue when I tried a different approach to do integration testing for the mongoose model, but it only affects test running.

@philippemiguet suggested that we remove those new tests from this pull request and open a different thread about it in-between cycles. I will be sending a detailed email soon to get feedback and ideas.

File: src/rpc/updateEmail.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  const user = await AuthenticationAccountModel.findOneByIdOrEmail({
+    _id,
+    email,
+  })
+  if (
+    !user ||
+    !(await user.updateEmailWithVerification({ password, newEmail }))
+  ) {
+    // intentionally generic error
```
Comment (philippemiguet) [inline src/rpc/updateEmail.js]: That would be my only thought: do we want to be generic at this level? I think that's great for now, we can revisit if we want a more accurate error in the future.

File: README.md [hunk 5]
```diff
@@ -305,7 +305,7 @@ Updates an existing accounts password (requires existing password)
 **Input**
 
 ```js
-rpc.call('removeProductlink', {
+rpc.call('updatePassword', {
```
Comment (philippemiguet) [inline README.md]: Hehe great spot!

**Comment by hharnisc** [comment]
Going to trigger a rebuild (via empty commit)

**Comment by hharnisc** [comment]
Hey @kiriappeee Not sure why this one is failing. Looks like Kuberdash is sending a 500 for some reason :thinking:

**Comment by philippemiguet** [comment]
@hharnisc @esclapes - From a convo I had last week with Adnan, we just need to merge master into that branch to make dev work again.

**Comment by esclapes** [comment]
OK. Just merged master as suggested by Adnan. Let's see... :crossed_fingers:

**Comment by hharnisc** [comment]
Hey @esclapes @philippemiguet I've updated the master branch and rebased it into this one. You'll need to stash any changes before doing a git pull on this branch.

**Comment by hharnisc** [comment]
Alright, looks like we're in a good place again! :tada:

**Pull request merged**